llvm/lib/Support/Mustache.cpp

```diff
@@ -681,10 +685,59 @@ void ASTNode::renderChild(const json::Value &Contexts, llvm::raw_ostream &OS) {
     Child->render(Contexts, OS);
 }
 
+void ASTNode::indentTextNode(std::string &Body, size_t Indentation,
+                             bool FinalNode) {
+  std::string spaces(Indentation, ' ');
+  size_t pos = 0;
+  size_t LastChar = std::string::npos;
+
+  if (FinalNode)
+    // body.erase(body.find_last_not_of(" \t\r\f\v") + 1);
+    LastChar = Body.find_last_not_of(" \t\r\f\v");
+
+  while ((pos = Body.find('\n', pos)) != std::string::npos) {
+    if ((!FinalNode) || (pos != LastChar)) {
+      Body.insert(pos + 1, spaces);
+      pos += 1 + Indentation;
+    } else {
+      break;
+    }
+  }
+}
+
+void ASTNode::indentNodes(ASTNode *Node, bool isPartial) {
+  size_t size = Node->Children.size();
+
+  for (size_t i = 0; i < size; ++i) {
+    ASTNode *child = Node->Children[i].get();
+    switch (child->Ty) {
+    case ASTNode::Text: {
+      indentTextNode(child->Body, Indentation, ((i == size - 1) && isPartial));
+      break;
+    }
+    case ASTNode::Section: {
+      indentNodes(child, false);
+      break;
+    }
+    case ASTNode::Partial: {
+      indentNodes(child, true);
+    }
+    case ASTNode::Root:
+    case ASTNode::Variable:
+    case ASTNode::UnescapeVariable:
+    case ASTNode::InvertSection:
+      break;
+    default:
+      llvm::outs() << "Invalid ASTNode type\n";
+      break;
+    }
```

Comment:
```suggestion
      llvm_unreachable("Invalid ASTNode type");
```
I believe we should get an error if we fail to handle one of the cases due to `-Wswitch-enum`, so default isn't needed. Since all cases are fully covered, I'd prefer we mark it unreachable, so it breaks loudly. If you're not getting the compile time warning or error diagnostic in your build for the uncovered switch, please let me know so we can fix that.

Reply:
I do get a warning about a fully covered switch with this, so we should remove it.